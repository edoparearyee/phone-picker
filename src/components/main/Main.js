import React, { Component } from 'react';

import './Main.css';
import Rating from '../rating/Rating';
import Options from '../options/Options';
import Price from '../price/Price';
import Option from '../option/Option';
import { load, loadFail, loadSuccess } from '../../actions';

/**
 * Main entry component for app
 *
 * @class Main
 * @extends {Component}
 */
class Main extends Component {
  /**
   * Load data from json file using fetch API and dispatch actions
   * when on success or failure
   */
  loadPhonesData() {
    const { dispatch } = this.props;
    dispatch(load());
    fetch('/data/phones.json')
      .then(res => res.json())
      .then(phones => dispatch(loadSuccess(phones)))
      .catch(() => dispatch(loadFail()));
  }

  /**
   * Fetch data once component is loaded in the DOM
   */
  componentDidMount() {
    this.loadPhonesData();
  }

  render() {
    const {
      phone,
      selectedColour,
      selectedMemory,
      selectedDevice,
      colourOptions,
      memoryOptions,
      loading,
      loadFailed,
      actions
    } = this.props;

    let content;

    if (loading) {
      // display loading message while data is requested
      content = <p className="Main__message">loadingâ€¦</p>;
    } else if (loadFailed) {
      // display error message if request fails
      content = (
        <p className="Main__message">
          Sorry there was an error loading this page
        </p>
      );
    } else {
      // display phone data once data is loaded
      content = (
        <article className="Product">
          <figure className="Product__gallery">
            <img
              className="Product__image"
              src={selectedDevice.merchandisingMedia[0].value}
              alt={selectedDevice.displayDescription}
            />
          </figure>

          <section className="Product__summary">
            <h1 className="Product__name">{phone.groupName}</h1>
            <Rating value={phone.rating} />
            <p className="Product__description">
              {selectedDevice.displayDescription}
            </p>

            <ul className="Product__options">
              <li className="Product__optionsItem Product__optionsColour">
                <Options label="Colour" value={selectedColour.colourName}>
                  {colourOptions.map(colour => (
                    <Option
                      active={selectedColour.colourHex === colour.colourHex}
                      color={colour.colourHex}
                      key={colour.colourHex}
                      value={colour}
                      action={actions.setColour}
                    >
                      <span className="Product__optionsColourName">
                        {colour.colourName}
                      </span>
                    </Option>
                  ))}
                </Options>
              </li>

              <li className="Product__optionsItem Product__optionsCapacity">
                <Options label="Capacity" value={selectedMemory}>
                  {memoryOptions.map(memory => (
                    <Option
                      active={selectedMemory === memory}
                      key={memory}
                      value={memory}
                      action={actions.setMemory}
                    >
                      {memory.slice(0, memory.length - 2)}
                    </Option>
                  ))}
                </Options>
              </li>
            </ul>

            <Price
              monthPrice={
                selectedDevice.priceInfo.bundlePrice.monthlyPrice.gross
              }
              upfrontPrice={
                selectedDevice.priceInfo.hardwarePrice.oneOffPrice.gross
              }
            />
          </section>
        </article>
      );
    }
    return (
      <main
        className={'Main' + (loading || loadFailed ? ' Main--hasMessage' : '')}
      >
        {content}
      </main>
    );
  }
}

export default Main;
