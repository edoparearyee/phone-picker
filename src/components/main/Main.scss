@import '../../styles/functions';
@import '../../styles/variables';

.Product {
  display: flex;
  padding: rem(40px);
}

.Product__summary,
.Product__gallery {
  flex: 1 0 50%;
  max-width: 50%;
}

.Product__gallery {
  align-items: flex-start;
  display: flex;
  justify-content: center;
}

.Product__image {
  max-width: 50%;
}

.Product__name {
  font-size: $font-size-h1;
  font-weight: 300;
  margin: 0 0 rem(30px);
}

.Product__description {
  margin-bottom: rem(50px);
}

.Product__options {
  display: flex;
  list-style: none;
  margin-bottom: rem(110px);
}

.Product__optionsItem {
  flex: 1 0 50%;
  max-width: 50%;
}

.Main {
  min-height: 100vh;
}

.Main--hasMessage {
  align-items: center;
  display: flex;
  justify-content: center;
}

.Main__message {
  text-align: center;
}
