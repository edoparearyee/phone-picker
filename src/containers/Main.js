import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import Main from '../components/main/Main';
import * as phonePickerActions from '../actions';
import {
  getMemoryOptions,
  getColourOptions,
  getLoading,
  getMemory,
  getSelectedDevice,
  getColourHex,
  getPhone,
  getLoadFailed
} from '../selectors';

const mapStateToProps = state => ({
  loading: getLoading(state),
  loadFailed: getLoadFailed(state),
  phone: getPhone(state),
  selectedColour: getColourHex(state),
  selectedMemory: getMemory(state),
  selectedDevice: getSelectedDevice(state),
  memoryOptions: getMemoryOptions(state),
  colourOptions: getColourOptions(state)
});

const mapDispatchToProps = dispatch => ({
  actions: bindActionCreators(phonePickerActions, dispatch),
  dispatch
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Main);
