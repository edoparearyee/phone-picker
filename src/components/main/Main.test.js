import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import * as phones from '../../phones.json';
import Main from './Main';

const phone = phones[0];

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <Main
      phone={phone}
      selectedDevice={phone.deviceSummary[0]}
      selectedColour={{ colourHex: '#f8e7dc', colourName: 'Gold' }}
      selectedMemory={'64GB'}
      memoryOptions={['64GB', '256GB']}
      colourOptions={[
        { colourHex: '#f8e7dc', colourName: 'Gold' },
        { colourHex: '#e1e2e3', colourName: 'Silver' },
        { colourHex: '#232324', colourName: 'Space Grey' }
      ]}
      actions={{
        setColour: value => console.log('setColour', value),
        setMemory: value => console.log('setMemory', value)
      }}
      dispatch={fn => null}
    />,
    div
  );
  ReactDOM.unmountComponentAtNode(div);
});

test('Main displays loading message', () => {
  const component = renderer.create(
    <Main
      loading={true}
      loadFailed={false}
      phone={null}
      selectedDevice={null}
      selectedColour={null}
      selectedMemory={null}
      memoryOptions={null}
      colourOptions={null}
      actions={{
        setColour: value => console.log('setColour', value),
        setMemory: value => console.log('setMemory', value)
      }}
      dispatch={fn => null}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Main displays phone information', () => {
  const component = renderer.create(
    <Main
      loading={false}
      loadFailed={false}
      phone={phone}
      selectedDevice={phone.deviceSummary[0]}
      selectedColour={{ colourHex: '#f8e7dc', colourName: 'Gold' }}
      selectedMemory={'64GB'}
      memoryOptions={['64GB', '256GB']}
      colourOptions={[
        { colourHex: '#f8e7dc', colourName: 'Gold' },
        { colourHex: '#e1e2e3', colourName: 'Silver' },
        { colourHex: '#232324', colourName: 'Space Grey' }
      ]}
      actions={{
        setColour: value => console.log('setColour', value),
        setMemory: value => console.log('setMemory', value)
      }}
      dispatch={fn => null}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Main displays error information', () => {
  const component = renderer.create(
    <Main
      loading={false}
      loadFailed={true}
      phone={null}
      selectedDevice={null}
      selectedColour={null}
      selectedMemory={null}
      memoryOptions={null}
      colourOptions={null}
      actions={{
        setColour: value => console.log('setColour', value),
        setMemory: value => console.log('setMemory', value)
      }}
      dispatch={fn => null}
    />
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
