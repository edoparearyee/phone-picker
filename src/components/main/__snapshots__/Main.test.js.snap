// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main displays error information 1`] = `
<main
  className="Main Main--hasMessage"
>
  <p
    className="Main__message"
  >
    Sorry there was an error loading this page
  </p>
</main>
`;

exports[`Main displays loading message 1`] = `
<main
  className="Main Main--hasMessage"
>
  <p
    className="Main__message"
  >
    loading…
  </p>
</main>
`;

exports[`Main displays phone information 1`] = `
<main
  className="Main"
>
  <article
    className="Product"
  >
    <figure
      className="Product__gallery"
    >
      <img
        alt="All-glass design, advanced cameras, wireless charging and a smart A11 Bionic chip. Intelligence never looked better."
        className="Product__image"
        src="/images/desktop/Apple_iPhone_8_Gold-full-product-front.png"
      />
    </figure>
    <section
      className="Product__summary"
    >
      <h1
        className="Product__name"
      >
        Apple iPhone 8
      </h1>
      <ul
        className="StarList"
      >
        <li
          className="StarList--item"
        >
          <svg
            className="Star Star--active"
          >
            <polygon
              points="9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78"
              style={
                Object {
                  "fillRule": "nonzero",
                }
              }
            />
          </svg>
        </li>
        <li
          className="StarList--item"
        >
          <svg
            className="Star Star--active"
          >
            <polygon
              points="9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78"
              style={
                Object {
                  "fillRule": "nonzero",
                }
              }
            />
          </svg>
        </li>
        <li
          className="StarList--item"
        >
          <svg
            className="Star Star--active"
          >
            <polygon
              points="9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78"
              style={
                Object {
                  "fillRule": "nonzero",
                }
              }
            />
          </svg>
        </li>
        <li
          className="StarList--item"
        >
          <svg
            className="Star Star--active"
          >
            <polygon
              points="9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78"
              style={
                Object {
                  "fillRule": "nonzero",
                }
              }
            />
          </svg>
        </li>
        <li
          className="StarList--item"
        >
          <svg
            className="Star Star--active"
          >
            <polygon
              points="9.9, 1.1, 3.3, 21.78, 19.8, 8.58, 0, 8.58, 16.5, 21.78"
              style={
                Object {
                  "fillRule": "nonzero",
                }
              }
            />
          </svg>
        </li>
      </ul>
      <p
        className="Product__description"
      >
        All-glass design, advanced cameras, wireless charging and a smart A11 Bionic chip. Intelligence never looked better.
      </p>
      <ul
        className="Product__options"
      >
        <li
          className="Product__optionsItem"
        >
          <div
            className="Options"
          >
            <p
              className="Options__state"
            >
              Colour
              : 
              <strong>
                Gold
              </strong>
            </p>
            <ul
              className="Options__list"
            >
              <li
                className="Option Option--active Option--color"
              >
                <button
                  className="Option__action"
                  onClick={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#f8e7dc",
                    }
                  }
                >
                  <span
                    className="Product__optionsColourName"
                  >
                    Gold
                  </span>
                </button>
              </li>
              <li
                className="Option Option--inactive Option--color"
              >
                <button
                  className="Option__action"
                  onClick={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#e1e2e3",
                    }
                  }
                >
                  <span
                    className="Product__optionsColourName"
                  >
                    Silver
                  </span>
                </button>
              </li>
              <li
                className="Option Option--inactive Option--color"
              >
                <button
                  className="Option__action"
                  onClick={[Function]}
                  style={
                    Object {
                      "backgroundColor": "#232324",
                    }
                  }
                >
                  <span
                    className="Product__optionsColourName"
                  >
                    Space Grey
                  </span>
                </button>
              </li>
            </ul>
          </div>
        </li>
        <li
          className="Product__optionsItem"
        >
          <div
            className="Options"
          >
            <p
              className="Options__state"
            >
              Capacity
              : 
              <strong>
                64GB
              </strong>
            </p>
            <ul
              className="Options__list"
            >
              <li
                className="Option Option--active "
              >
                <button
                  className="Option__action"
                  onClick={[Function]}
                  style={
                    Object {
                      "backgroundColor": undefined,
                    }
                  }
                >
                  64
                </button>
              </li>
              <li
                className="Option Option--inactive "
              >
                <button
                  className="Option__action"
                  onClick={[Function]}
                  style={
                    Object {
                      "backgroundColor": undefined,
                    }
                  }
                >
                  256
                </button>
              </li>
            </ul>
          </div>
        </li>
      </ul>
      <div
        className="Price"
      >
        <div
          className="Price__upfront"
        >
          <p>
            from
             
            <span
              className="Price__value"
            >
              £
              149
            </span>
             
            upfront
          </p>
        </div>
        <div
          className="Price__month"
        >
          <p>
            when you pay
             
            <span
              className="Price__value"
            >
              £
              43.20
            </span>
             a month
          </p>
        </div>
      </div>
    </section>
  </article>
</main>
`;
